TimeRangePicker: Improve accessibility

tskarhed opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Improve keyboard navigation and screen reader support
- Add landmark markup
- Add input markup to the TimeRangeList
- Add keyboard focus style to TimeRangeList
- Improve TimePickerCalendar keyboard navigation
- Creates narrow screen accessible accordion

Does not include improvements to TimeZonePicker or the Calendar.

**Which issue(s) this PR fixes**:
Relates to #35613

tskarhed committed:
Add low hanging aria markup

tskarhed committed:
Add a list of radio buttons

tskarhed committed:
Add ref to put focus

tskarhed committed:
Add landmark markup

tskarhed committed:
Add labels to icon buttons

tskarhed committed:
Add radio button groupings

tskarhed committed:
Merge branch 'main' into a11y-timepicker

tskarhed committed:
Add keyboard focus styles

tskarhed committed:
Improve markup

tskarhed commented:
It seems like a scrollbar appears in some place here. Not sure why.
![Screenshot from 2021-07-22 11-41-23](https://user-images.githubusercontent.com/1438972/126619986-b8fa919f-b57a-4bc5-a27d-849eb328ec24.png)

tskarhed committed:
Update test selector

tskarhed committed:
Remove testid in favor of text

tskarhed committed:
Fiddle around with test labels

tskarhed committed:
Get correct testid selector

tskarhed commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerContent.tsx:
```diff
@@ -218,19 +229,27 @@ const NarrowScreenForm: React.FC<FormProps> = (props) => {
-        <TimePickerTitle>Absolute time range</TimePickerTitle>
-        {!hideQuickRanges && <Icon name={!collapsed ? 'angle-up' : 'angle-down'} />}
-      </div>
+        <button
```
Needs an `aria-expanded` attribute

tskarhed committed:
Add aria-expanded to button

tskarhed committed:
Change order of quick select for keyboard nav

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=36912) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=36912) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: tskarhed<br/>:x: Tobias Skarhed<br/><hr/>**Tobias Skarhed** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=36912) it.</sub>

tskarhed commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx:
```diff
@@ -16,26 +18,52 @@ const getStyles = (theme: GrafanaTheme2) => {
+        className={styles.radio}
+        checked={selected}
+        name={name}
+        type="radio"
```
Radio groups causes automatic selection when navigating with keyboard - should we change to checkboxes or override the default behaviour?

thisisobate replied:
I noticed some behavior with the checkboxes when interacting using the screen reader. The screen reader does not identify a `checked` box immediately it's been checked. It reads it as `'not checked'`. But then, clicking the TimePicker again shows the selected option as `checked`. 

I assume this is an issue with the screen reader software. Have you encountered anything like that? @tskarhed

tskarhed committed:
Create headings for easier heading navigation

tskarhed committed:
Add another h3

tskarhed commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerContent.tsx:
```diff
@@ -218,21 +230,24 @@ const NarrowScreenForm: React.FC<FormProps> = (props) => {
-        {!hideQuickRanges && <Icon name={!collapsed ? 'angle-up' : 'angle-down'} />}
-      </div>
+    <fieldset>
+      <h3 className={styles.header}>
```
Used an h3 here based on https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html

tskarhed committed:
Add testid

tskarhed committed:
Use selectors in e2e test

tskarhed committed:
Add dataid selector for narrow absolute button

tskarhed committed:
Update e2e test with proper selectors

thisisobate commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.tsx:
```diff
@@ -89,11 +90,21 @@ export class UnthemedTimeRangePicker extends PureComponent<TimeRangePickerProps,
           <ToolbarButton
-            aria-label="TimePicker Open Button"
+            data-testid={selectors.components.TimePicker.openButton}
+            aria-label={`Time range editor ${formattedRange(value, timeZone)}`}
```
This works great! 🎉👍

thisisobate replied:
I feel we need to use a lot more descriptive text in the Time range editor label text. Currently, the screen reader reads it as "Time range editor {any time range selected}" which might be a bit difficult for the user to comprehend. 

Rather, we could use something like `"Time range editor with current time range {any time range selected}"`

thisisobate commented on packages/grafana-e2e-selectors/src/selectors/components.ts:
```diff
@@ -6,11 +6,12 @@
 export const Components = {
   TimePicker: {
-    openButton: 'TimePicker Open Button',
+    openButton: 'data-testidTimePicker Open Button',
```
`'data-testid TimePicker Open Button'`

thisisobate reviewed (changes requested):
Amazing work!  @tskarhed 🎉🎉

Left some thoughts for us to consider.

tskarhed commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerFooter.tsx:
```diff
@@ -59,13 +59,13 @@ export const TimePickerFooter: FC<Props> = (props) => {
 
   return (
-    <div className={style.container}>
+    <section aria-label="Time zone selection" className={style.container}>
```
Tab navigation is usually done for components you can interact with. The panel header button should be enough for that. Adding section causes the screen reader to announce that you're entering or leaving a new region (with the aria label as its name). Using screen readers you can also do shortcuts between all the sections on the page.

The components inside the section are possible to navigate with a keyboard, so no reason why you would want to add another step of a non-interactable element.

tskarhed committed:
Add more verbose aria-label

tskarhed committed:
Merge branch 'main' into a11y-timepicker

thisisobate reviewed (approved):
LGTM! 🎉

kaydelaney commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.test.tsx:
```diff
@@ -26,6 +26,6 @@ describe('TimePicker', () => {
     );
 
-    expect(container.queryByLabelText(/timepicker open button/i)).toBeInTheDocument();
+    expect(container.queryByLabelText(/Time range editor/i)).toBeInTheDocument();
```
Should we go with "Time range editor" or "Time range picker" here? Just want to make sure whatever it is, it's consistent.

tskarhed replied:
Good point, I'll change it to TimeRange picker instead

kaydelaney commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx:
```diff
@@ -16,26 +18,52 @@ const getStyles = (theme: GrafanaTheme2) => {
+export const TimeRangeOption = memo<Props>(({ value, onSelect, selected = false, name }) => {
   const styles = useStyles2(getStyles);
+  // In case there are more of the same timerange in the list
+  const id = value.display + Math.floor(Math.random() * 10);
```
Could also use the `uuid` package

kaydelaney reviewed (approved):
Really awesome work here Tobias! I think this is a good start for the time range picker.
Some thoughts:
In Firefox it seems to take two tab presses to get from the time-picker button to highlighting the first relative time range (chrome seems to only need one).

It’s maybe a little surprising/confusing that the tab order starts with the relative time ranges - maybe starting with the “From” input would make more sense? Let me know what you think.

Also I’m not 100% sure about having all the relative time ranges be tab-navigable, just because having to go through all of them to get focus on the next toolbar item could be a little tedious, maybe the arrow keys could work better for those?
Lastly it might be nice to be able to press “Esc” and have the calendar close (not sure if role=“menu” semantics are applicable here, but something to look into).

Maybe you've read it before, but I've found https://w3c.github.io/aria-practices/#keyboard to be a big help.

In any case, awesome work. 👍

tskarhed commented:
It's a bit frustrating with the way non-screen reader keyboard navigation handles radio buttons, which is the reason I changed to checkboxes. Then you're able to actually change the selection. The problem with radio buttons is also that the default and recommended behaviour is that the selection changes when moving with the arrow keys, causing the popup to close.

I changed so that the relative time ranges would come first in the order, but this is probably just a personal preference. The best thing we could do here is to ask a screen reader user.

tskarhed committed:
Change editor to picker
